### PR TITLE
amclient: add backoff to TransferSession.Start()

### PR DESCRIPTION
**Pull request by sevein**
Introduce retry with exponential back-off in amclient's TransferSession.Start().

Typically I wouldn't vendor a new package for a simple retry but we wanted to use an exponential back-off algorithm and that became much harder so I ended up using [github.com/cenkalti/backoff](https://github.com/cenkalti/backoff) which gives us a stable and configurable implementation that we can also use in other situations, e.g. http/s3 downloads, etc...

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -6,7 +6,7 @@ VERSION := $(shell git describe --tags --always --dirty)
 default: testrace vet
 
 tools:
-	go get -u github.com/golang/dep/...
+	go get -u github.com/golang/dep/cmd/...
```
Comment (sevein) [inline Makefile]: I'm trying to squeeze this commit in this PR so I don't have to submit another PR. `make tools` is just a way to download all the Go tools that the developer needs. The import path used to download the dep package management tool wasn't correct and this is fixing it.

**Comment by sevein** [comment]
I've made some tests and it seems to be working properly. I've sent messages to the adapter while the dashboard was down. The adapter waits up to 30 minutes before it gives up and the maximum interval between attempts is 30 seconds. These values are hard-coded for now but they're very generous.

File: amclient/transfer_session.go [hunk 2]
```diff
@@ -135,41 +119,41 @@ func (s *TransferSession) Start() error {
...
 		for _, item := range payload.Results {
 			if item.Directory == basePath {
 				_, _, err := s.c.Transfer.Approve(ctx, &TransferApproveRequest{
 					Directory: s.Path,
 					Type:      "standard",
 				})
 				if err != nil {
-					return errors.Wrap(err, "approve request failed")
+					// We've tried but it failed. Keep trying.
+					return backoff.Permanent(errors.Wrap(err, "approve request failed"))
```
Comment (lower29) [inline amclient/transfer_session.go]: The comment here appears to go against what the contract of `backoff.Permanent` describes. The [GoDoc](https://godoc.org/github.com/cenkalti/backoff#Retry) states that if `Operation` returns a `PermanentError` (which is what `Permanent` returns) then the operation **will not** be reattempted by the `Retry` function.

So which of these was the intent? Do you want the operation to continue to be retried if the server returns an error, or do you want it to abort if the server returns an error? I think either the comment needs changing or the code does.

**Comment by helenst** [comment]
Just saw @lower29 also picked up on that comment that didn't seem right :smile:

One more thing @sevein - you noted in the comment above that it will wait 30 minutes before giving up, but in the code it's 10 minutes. Did you mean to have that value in the code (or was it a typo in the comment)?

File: amclient/transfer_session.go (hunk 2 above)
Comment (sevein) [reply amclient/transfer_session.go]: Firstly I thought I wouldn't retry if I get an actual HTTP error but then I realized it would be good to retry under these circumstances too. Thanks, I'm glad you guys are paying attention!

**Comment by sevein** [comment]
Thank you so much. I meant 10 minutes :) And I didn't want to leave that `PermanentError`. After some thinking I couldn't find a situation where I really wanted to abort the operation, even in case of errors or if the transfer is not listed yet I think it's important to keep trying.

**Comment by lower29** [comment]
Changes in response to our comments look good to me 👍
